libraries/AC_Avoidance/AP_OADijkstra.cpp

```diff
@@ -0,0 +1,550 @@
+// resulting path is stored in _shortest_path array as vector offsets from EKF origin
+bool AP_OADijkstra::calc_shortest_path(const Location &origin, const Location &destination)
+{
+    // convert origin and destination to offsets from EKF origin
+    Vector2f origin_NE, destination_NE;
+    if (!origin.get_vector_xy_from_origin_NE(origin_NE) || !destination.get_vector_xy_from_origin_NE(destination_NE)) {
+        return false;
+    }
+
+    // create origin and destination visgraphs of polygon points
+    update_visgraph(_source_visgraph, {AP_OAVisGraph::OATYPE_SOURCE, 0}, origin_NE, true, destination_NE);
+    update_visgraph(_destination_visgraph, {AP_OAVisGraph::OATYPE_DESTINATION, 0}, destination_NE);
+
+    // clear short_path_data array
+    _short_path_data_numpoints = 0;
+
+    // add origin and destination (node_type, id, visited, distance, distance_from_idx) to short_path_data array
+    _short_path_data[_short_path_data_numpoints] = {{AP_OAVisGraph::OATYPE_SOURCE, 0}, false, 0, 0};
+    _short_path_data_numpoints++;
+    _short_path_data[_short_path_data_numpoints] = {{AP_OAVisGraph::OATYPE_DESTINATION, 0}, false, OA_DIJKSTRA_POLYGON_VISGRAPH_INFINITY_CM, OA_DIJKSTRA_POLYGON_SHORTPATH_NOTSET_IDX};
+    _short_path_data_numpoints++;
+
+    // add fence points to short_path_data array (node_type, id, visited, distance, distance_from_idx)
+    // skip last fence point because it is the same as the first
+    for (uint8_t i=0; i<_polyfence_numpoints-1; i++) {
+        _short_path_data[_short_path_data_numpoints] = {{AP_OAVisGraph::OATYPE_FENCE_POINT, i}, false, OA_DIJKSTRA_POLYGON_VISGRAPH_INFINITY_CM, OA_DIJKSTRA_POLYGON_SHORTPATH_NOTSET_IDX};
+        _short_path_data_numpoints++;
+    }
+
+    // start algorithm from source point
+    node_index current_node_idx = 0;
+
+    // update nodes visible from source point
+    for (uint8_t i=0; i<_source_visgraph.num_items(); i++) {
+        node_index node_idx;
+        if (find_node_from_id(_source_visgraph[i].id2, node_idx)) {
+            _short_path_data[node_idx].distance_cm = _source_visgraph[i].distance_cm;
+            _short_path_data[node_idx].distance_from_idx = current_node_idx;
+        } else {
+            return false;
+        }
+    }
+    // mark source node as visited
+    _short_path_data[current_node_idx].visited = true;
+
+    // move current_node_idx to node with lowest distance
+    while (find_closest_node_idx(current_node_idx)) {
+        // update distances to all neighbours of current node
+        update_visible_node_distances(current_node_idx);
+
+        // mark current node as visited
+        _short_path_data[current_node_idx].visited = true;
+    }
+
+    // extract path starting from destination
+    bool success = false;
+    {
+        node_index nidx;
+        if (!find_node_from_id({AP_OAVisGraph::OATYPE_DESTINATION,0}, nidx)) {
+            return false;
+        }
+        _path_numpoints = 0;
+        bool complete = false;
```

Comment:
You don't need the complete variable, it actually makes it more complicated (and slower). Just make this a while (true) and break once you get to the complete case.

Reply:
ok, I feel odd about having a, "while (true)" in there but I'll change it txs..